src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Windows.cs

```diff
@@ -286,12 +287,36 @@ unsafe long GetFileLengthCore()
             {
                 Interop.Kernel32.FILE_STANDARD_INFO info;
 
-                if (!Interop.Kernel32.GetFileInformationByHandleEx(this, Interop.Kernel32.FileStandardInfo, &info, (uint)sizeof(Interop.Kernel32.FILE_STANDARD_INFO)))
+                if (Interop.Kernel32.GetFileInformationByHandleEx(this, Interop.Kernel32.FileStandardInfo, &info, (uint)sizeof(Interop.Kernel32.FILE_STANDARD_INFO)))
+                {
+                    return info.EndOfFile;
+                }
+
+                if (_path is null || !PathInternal.IsDevice(_path))
+                {
+                    throw Win32Marshal.GetExceptionForLastWin32Error(Path);
+                }
+
+                const int requiredBufferSize = 32;
+                byte* pBuffer = stackalloc byte[requiredBufferSize];
```

Comment:
Instead of this, define `STORAGE_READ_CAPACITY storageReadCapacity` local variable and pass pointer to it directly to `DeviceIoControl`.